scripts/firedrake-install

```diff
@@ -0,0 +1,395 @@
+        if hasattr(sys, "real_prefix"):
+            check_call(["pip", "install", "virtualenv"])
+        else:
+            check_call(["pip", "install", "--user", "virtualenv"])
+        quit("Virtual env installed. Please run firedrake-install again.\n")
+    if mode == "install":
+        stdout.write("Creating firedrake virtualenv.\n")
+        virtualenv.create_environment("firedrake", site_packages=True)
+    execfile("firedrake/bin/activate_this.py", dict(__file__='path/to/activate_this.py'))
+
+os.chdir("firedrake")
+if mode == "install":
+    os.mkdir("src")
+    os.chdir("src")
+
+    git_clone("git+https://github.com/firedrakeproject/firedrake.git@install_script")
+    packages = clone_dependencies("firedrake")
+    packages = clone_dependencies("PyOP2") + packages
+    packages += ["firedrake"]
+
+    for p in packages:
+        pip_requirements(p)
+
+    sudopip.append("--no-deps")
+    for p in packages:
+        install(p+"/")
+        if args.developer:
+            pdir = os.getcwd() + "/" + p + os.pathsep
+            os.environ["PYTHONPATH"] = pdir + os.environ.get("PYTHONPATH", "")
+            path_extension = pdir + path_extension
+
+    # Work around easy-install.pth bug.
+    if args.developer:
+        packages.remove("petsc")
+        packages.remove("petsc4py")
+        packages.remove("firedrake")
+        v = sys.version_info
+        easyinstall = file("../lib/python%d.%d/site-packages/easy-install.pth" % (v.major, v.minor), "r").readlines()
+        new_packages = [os.getcwd() + "/" + p + "\n" for p in packages]
+        file("../lib/python%d.%d/site-packages/easy-install.pth" % (v.major, v.minor), "w").writelines(
+            easyinstall[:1] + new_packages + easyinstall[1:])
+
+else:
+    # Update mode
+    os.chdir("src")
+    packages = os.listdir(".")
+    # update packages.
+    petsc_changed = git_update("petsc")
+    petsc4py_changed = git_update("petsc4py")
+
+    packages.remove("petsc")
+    packages.remove("petsc4py")
+
+    for p in packages:
+        git_update(p)
+
+    # update dependencies.
+    pip_requirements("PyOP2")
+    pip_requirements("firedrake")
+    sudopip.append("--no-deps")
+
+    # Only rebuild petsc if it has changed.
+    if args.rebuild or petsc_changed:
+        clean("petsc/")
+        install("petsc/")
+    if args.rebuild or petsc_changed or petsc4py_changed:
+        clean("petsc4py/")
+        install("petsc4py/")
+    packages.remove("PyOP2")
+    packages.remove("firedrake")
+    packages += ("PyOP2", "firedrake")
+    for p in packages:
+        clean(p)
+        install(p)
+
+stdout.write("Creating firedrake-update script.\n")
+update_script = open("firedrake/scripts/firedrake-install", "r").read()
+
+for switch in ["developer", "user", "system", "sudo", "prefix"]:
+    update_script = update_script.replace("args.%s = False" % switch,
+                                          "args.%s = %s" % (switch, args.__getattribute__(switch)))
```

Comment:
I actually think this is the least bad option. Basically if you install one way and then try to upgrade another way you are going to break things. It's therefore safer to not offer the option.